Add functions to get 3D krypton maps

mcidlaso opened the pull request:
These scripts contain functions to:  
- Make cuts on the outputs from sophronia to get the selected events and the efficiencies of the cuts
- Get the lifetime and drift velocity for each run 
- Compute 3D krypton maps and apply them to correct the data
- Some tests for these functions (still have a lot of work to do)

mcidlaso committed:
Add krmap functions

mcidlaso committed:
Add correction functions

mcidlaso committed:
Add selection functions

mcidlaso committed:
Add test functions for correction functions

mcidlaso committed:
Add lifetime fit and drift velocity functions

mcidlaso committed:
Add tests functions for the krmap functions

mcidlaso committed:
Add tests functions for the krmap functions

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,51 @@
+
+
+
+def normalization(krmap, method, x_low, x_high, y_low, y_high):
```
The last 4 parameters need to be optional, since they are only relevant for `method == "region"`. I also think it would be interesting to gather them in a dictionary to simplify the logic.

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,51 @@
+
+def normalization(krmap, method, x_low, x_high, y_low, y_high):
+
+    mu_values = krmap.mu
```
since you are doing `.dropna()` everywhere, you can do it here already

mcidlaso committed:
correct error

mcidlaso committed:
correct error

gonzaponte commented on invisible_cities/icaros/correction_functions_test.py:
```diff
@@ -0,0 +1,43 @@
+    k, i, j = np.meshgrid(k_vals, i_vals, j_vals, indexing='ij')
+    k = k.ravel()
+    i = i.ravel()
+    j = j.ravel()
```
```suggestion
    k, i, j = map(np.ravel,
                  np.meshgrid(k_vals, i_vals, j_vals, indexing='ij'))
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+"""
+
+def create_NaN_map(xy_range, dt_range, xy_nbins, dt_nbins):
```
```suggestion
def create_empty_map(xy_range, dt_range, xy_nbins, dt_nbins):
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+    i_range = np.arange(0, len(xy_bins)-1)
+    j_range = np.arange(0, len(xy_bins)-1)
+    k_range = np.arange(0, len(dt_bins)-1)
```
`xy_bins` and `dt_bins` are only used to compute their length, which are given by the variables `xy_nbins` and `dt_nbins`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+
+def med_fun(df):
+    sigma = (0.04/2.35)*df.S2e.median()
```
Add a comment explaining this value

mcidlaso committed:
Correct error

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    return Nan_map
+
+
+def med_fun(df):
```
rename to `median`, `get_median` or similar

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+
+
+def fit_function(df, bins, size =10):
```
rename to `gaussian_fit` or similar
rename `bins` to `ebins`, `energy_bins`, or similar, to clarify that they refer to energy
rename `size` to something that gives a better hint of its meaning (e.g. `min_events`)

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    if len(df)<size:
+        results = med_fun(df)
+
+    bin_centers = shift_to_bin_centers(bin_edges)
```
if `bin_edges` is only used to compute `bin_centers`, then the argument to this function should be `bin_centers`

mcidlaso committed:
Correct error again

mcidlaso committed:
correct all errors

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    return results
+
+
+def map_3D_fits(df, xy_range, dt_range, xy_nbins, dt_nbins):
```
try to come up with a name like `verb_noun`, like `create_something`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+    x = df.X
+    y = df.Y
+    dt = df.DT
```
```suggestion
    x, y, dt, _ = df.values.T
```
and then you don't need to do .values on each column

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+
+    map_df = df.assign( i = np.digitize(x.values, bins = xy_bins) - 1,
+                        j = np.digitize(y.values, bins = xy_bins) - 1,
+                        k = np.digitize(dt.values, bins = dt_bins) - 1)
```
Add a comment explaining the -1 at the end

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+                        k = np.digitize(dt.values, bins = dt_bins) - 1)
+
+
+    result = map_df.groupby(['k', 'i', 'j']).apply(fit_function, bins = np.linspace(3000, 9000, 101))
```
the argument bins should probably be an argument to this function as well

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    result = result.reset_index()
+
+    if 'level_3' in result.columns:
+        result = result.drop(columns = 'level_3')
```
This should be deterministic, either the condition is true or false, but it will not change depending on the data, I don't think

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+def merge_maps(NaN_map, map_3D):
+
+    full_map = pd.concat([map_3D, NaN_map], ignore_index = True)
+    full_map = full_map.groupby(['k', 'i','j']).first().reset_index()
```
Explain this bit in a comment

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    y = shift_to_bin_centers(xy_bins)
+    dt = shift_to_bin_centers(dt_bins)
+
+    combinaciones_bins = pd.DataFrame(
```
english :)

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,149 @@
+    columns=['dt', 'x', 'y'])
+
+
+    map = pd.concat([combinaciones_bins.reset_index(drop=True), map.reset_index(drop=True)], axis = 1)
```
why do you need to reset indices here?

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,85 @@
+    assert result.shape == (1, 5)
+    assert result['nevents'].sum() == 0
+    for col in ['median', 'sigma', 'median_error', 'sigma_error']:
+        assert pd.isna(result[col].iloc[0])
```
```suggestion
        assert pd.isna(result[col].iloc[0]), f"{col} is not a nan"
```

this adds a bit of information in case of failure

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,85 @@
+
+
+
+def test_empty_medfun_2():
```
```suggestion
def test_medfun_empty_input_does_not_raise():
```

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,85 @@
+from pytest import raises
+
+
+def test_empty_medfun():
```
Try to be a bit more descriptive in the test name. Also keep the structure `test_TESTEDFUN_DESCRIPTION`:

```suggestion
def test_medfun_empty_input():
```

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def test_Nevents():
+    d = {'DT': np.empty(6, dtype=int), 'x' : np.empty(6), 'y' : np.empty(6),'S2e' : np.ones(6)}
```
why `dtype=int` here?

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,85 @@
+    map_3D_test = pd.DataFrame(data = d, index = range(0, 6))
+    assert merge_maps(Nan_map_test, map_3D_test).shape == Nan_map_test.shape
+
+test_merge_maps()
```
remove this line

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,137 @@
+
+
+
+def select_lifetime_region(df, bins, x0, y0, r, half_size, circle=True):
```
we should review this function interactively

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,137 @@
+def LT_fit(DT, s2e, p0):
+
+    def exponential(t, e, tau):
+        return e*np.exp(-t/tau)
```
Use `expo` from `fit_functions` (careful with the negative sign)

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,137 @@
+    def exponential(t, e, tau):
+        return e*np.exp(-t/tau)
+
+    popt, pcov = curve_fit(exponential, DT, s2e, p0 = p0)
```
Use `fit` from `fit_functions`

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,137 @@
+        Array of computed sigmoid values for x array.
+    '''
+    sigmoid = scale / (1 + np.exp(-slope * (x - inflection))) + offset
+    return sigmoid
```
we can move this function to `fit_functions`

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+    if do_print:
+        seff = "{:.4f}".format(eff)
+        print(f"{name} efficiency {seff}, {nsel}/{ntot}.")
+    return eff
```
I think this function is replaced by `eff_of_selection`, isn't it?

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+        file = [os.path.join(path, f) for f in os.listdir(path) if f.endswith(".h5")]
+        filenames.append(file)
+    filenames = list(itertools.chain.from_iterable(filenames))
+    return filenames
```
This function needs to be more generic, instead of assuming that the folder is structured like
```
top
└── trigger1
    ├── ldc1
    ├── ldc2
    ├── ldc3
    ├── ldc4
    ├── ldc5
    ├── ldc6
    └── ldc7
```

Maybe it can search for all `.h5` files recursively

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+    eff = events_after/events_before
+
+    print(f"{name} efficiency {eff}, {events_after}/{events_before}.")
```
Wrap it in `if name:`?

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+def select_Xrays(kdst, low_xrays, high_xrays):
+
+    sel_xrays = (kdst.Ec >= low_xrays) & (kdst.Ec <= high_xrays)
```
use `in_range`

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+def select_1S1_1S2(kdst):
+    a = kdst.groupby('event')['time'].count()
+    events_1S1_1S2 = a[a == 1].index.values
```
explain this and rename `a` to something more self-explanatory

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+def select_S2t(kdst, low_S2t, high_S2t):
+    range_S2t = (low_S2t, high_S2t)
+    sel_S2t = in_range(kdst.S2t, *range_S2t)
```
```suggestion
    sel_S2t = in_range(kdst.S2t, low_S2t, high_S2t)
```

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+
+def select_DTrange(kdst, low_DT, high_DT):
+    df_DTrange = kdst[(kdst.DT >= low_DT) & (kdst.DT <= high_DT)]
```
Use `in_range`

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+
+def select_nsipm(kdst, low_nsipm, high_nsipm):
+    sel_nsipm = (kdst.Nsipm >= low_nsipm) & (kdst.Nsipm <= high_nsipm)
```
Use `in_range`

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+
+    df, eff_1S1_1S2 = select_1S1_1S2(df)
+
+    df, eff_S2t = select_S2t(df, low_S2t, high_S2t)
```
This one should go before the 1S1+1S2 selection, I think

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,148 @@
+         'eff S2 trigger time': [eff_S2t], 'eff Rmax': [eff_Rmax], 'eff range DT': [eff_DTrange],
+         'eff number of SiPMS': [eff_nsipm], 'total efficiency': [total_eff]}
+
+    df_efficiencies = pd.DataFrame(data = d)
```
You can also have numbers in the dictionary (instead of lists of 1 element) and then add `index=[0]` when creating the dataframe

mcidlaso committed:
Drop dropna() from normalization function

mcidlaso committed:
Add optional parameters in a single argument

mcidlaso committed:
Import functions for creating krmaps and t_evol in the same file

mcidlaso committed:
Change description and correct create_empty_map function

mcidlaso committed:
Change name and add explanation comment in get_median() function

mcidlaso committed:
Change function and variables names in gaussian_fit function

mcidlaso committed:
Change function name from map_3D_fits to fit_map

mcidlaso committed:
Changes in fit_map function

mcidlaso committed:
Add description for merge_maps function

mcidlaso committed:
Change variable name for krmap in include_coordinates function

mcidlaso committed:
Change variable and function names in compute_3D_map function

mcidlaso committed:
Add function to get map info (metadata)

mcidlaso committed:
Add functions (from icaros/krcal/core/fit_functions.py) to do fits

mcidlaso committed:
Add function to divide dataframe into time slices

mcidlaso committed:
Add function to calculate time evolution parameters

mcidlaso committed:
Add function to save map

mcidlaso committed:
Simplify select_lifetime_region function

mcidlaso committed:
Change LT_fit function to use fit function from IC

mcidlaso committed:
Comment print

mcidlaso committed:
Change and move eff_of_sel function

mcidlaso committed:
Change load_files to be more straightforward

mcidlaso committed:
Move and change line in eff_of_selection function

mcidlaso committed:
Add line in apply_correctionmap function so the corrected kdst ends up with two columns for corrected energies

mcidlaso committed:
Use in_range function from IC

mcidlaso committed:
Add comment explaining the function and change variable name

mcidlaso committed:
Remove range_S2t variable

mcidlaso committed:
Use in_range function from IC

mcidlaso committed:
Change order on selections and correct typo

mcidlaso committed:
Change column names so we don't get a warning when using df_writer

mcidlaso committed:
Create test for lifetime and drift velocity

mcidlaso committed:
Change test names to be more explicit

mcidlaso committed:
Change function names, remove one line and remove dtype = int

mcidlaso committed:
Remove comment

mcidlaso committed:
Change function that computes metadata

mcidlaso committed:
Change function that saves maps so it doesn't raise an error when opening metadata

mcidlaso committed:
Monitoring function to plot time evolution parameters

mcidlaso committed:
Change function so the yranges of the plots depend on the values of the parameters

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,53 @@
+
+def normalization(krmap, method, xy_params = None):
+
+    mu_values = krmap.mu.dropna()
```
`dropna` should affect all options of method. Right now, it's only affecting the first two, the other three do not use this variable. You should redefine kr_map so it doesn't contain `nan` values.:
`krmap = krmap.dropna(subset=["mu"])`. Then use `krmap.mu.something` everywhere

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,53 @@
+from scipy.interpolate import griddata
+
+
+def normalization(krmap, method, xy_params = None):
```
We will discuss how to do this with symbols

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,53 @@
+        E_reference_anode = mu_values_anode.mean()
+        return E_reference_anode
+
+    mask_region = (krmap['x'] <= xy_params['x_high']) & (krmap['x'] >= xy_params['x_low']) & (krmap['y'] <= xy_params['y_high']) & (krmap['y'] >= xy_params['y_low'])
```
use `in_range`

gonzaponte commented on invisible_cities/icaros/correction_functions_test.py:
```diff
@@ -0,0 +1,43 @@
+        (map_test.x <= 100) & (map_test.x >= -100) &
+        (map_test.y <= 100) & (map_test.y >= -100)
+    )
+
```
`in_range`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+
+def gaussian_fit(df, ebins, min_events = 10):
+    if len(df)< min_events:
+        results = get_median(df)
```
```suggestion
        return get_median(df)
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        'method'      : norm_method,
+        'dtbins'      : [np.unique(krmap.k.values).tolist()],
+        'xbins'       : [np.unique(krmap.i.values).tolist()],
+        'ybins'       : [np.unique(krmap.j.values).tolist()],
```
```suggestion
        'dtbins'      : [krmap.k.unique().to_list()],
        'xbins'       : [krmap.i.unique().to_list()],
        'ybins'       : [krmap.j.unique().to_list()],
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        sigma = poisson_sigma(y)
+        f     = fit(gauss, x, y, seed, sigma = sigma)
+    else:
+        f     = fit(gauss, x, y, seed)
```
```suggestion
    sigma = poisson_sigma(y) if sigma else None
    f     = fit(gauss, x, y, seed, sigma = sigma)
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        f     = fit(gauss, x, y, seed, sigma = sigma)
+    else:
+        f     = fit(gauss, x, y, seed)
+    assert np.all(f.values != seed)
```
```suggestion
    if not np.all(f.values != seed):
        raise RuntimeError("a meaningful error message")
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        t_start = time_edges[i]
+        t_end   = time_edges[i + 1]
+        
+        mask = (df.time >= t_start) & (df.time < t_end)
```
`in_range`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+    
+    df_tevols = []
+    
+    for i in range(len(df)):
```
rename `df` to `dfs` and then do `for df in dfs:`. With this you don't need to put `dfs[i]` everywhere, but simply `df`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        e0 = magnitudes[0]
+        e0u = uncertainties[0]
+    
+        dv, udv = compute_drift_v(df[i].DT.to_numpy(), df[i].DT.nunique(), dtrange, seed)
```
why is the number of bins set to the number of unique dt values?

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        u_mu = f.errors[1]
+    
+        sigma = f.values[2]
+        u_sigma = f.errors[2]
```
```suggestion
        _,   mu,   sigma = f.values
        _, u_mu, u_sigma = f.errors
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+        R = np.sqrt(8*np.log(2))*(sigma/mu) #FWHM (krypton paper)
+        u_R = R * (u_mu**2/mu**2 + u_sigma**2/sigma**2)**0.5
+    
+        mask = (df[i].DT.values > low_DT) & (df[i].DT.values < high_DT)
```
`in_range`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+    for i in range(len(df)):
+        
+        mean = df[i].S2e.mean()
+        std = df[i].S2e.std()/np.sqrt(len(df[i]))
```
define `sqrtn = np.sqrt(len(df[i]))`  and use this variable for the rest of the function to reduce the clutter

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,316 @@
+  
+
+
+def get_time_evol(df, run_number, bins_lt, x0, y0, shape, shape_size, p0, dtrange, low_DT, high_DT, bins_Ec, sigma = False, seed = None): 
```
let's talk about how to refactor this using `groupby`

mcidlaso committed:
Import functions with the right name

mcidlaso committed:
Import functions to make tests and change one line in test_method_norm()

mcidlaso committed:
Add test for apply_3Dmap function

mcidlaso committed:
Add one comment

mcidlaso committed:
Add control plots functions

mcidlaso committed:
Added run numbers, bins for each variable and put ylimits in one plot

mcidlaso committed:
Added normalization methods to calculate the median values of mu

mcidlaso committed:
Corrected name error in get_time_evol function and now the resolution is correct

mcidlaso committed:
Added functions to append multiple time evolutions from different runs for monitoring

mcidlaso committed:
Removed run_number variable because it wasnt being used

mcidlaso committed:
Correct typo

mcidlaso committed:
Include median in normalization function test

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    nevents = len(df['event'].unique())
+    
+    df2 = df.groupby('event s1_peak'.split()).first().reset_index()
+    df2_ = df2.groupby('event').count()
```
I think you don't need to define a new variable here and reuse df2. Same for `df1` below

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    axs[0, 1].legend()
+
+
+    axs[1, 0].hist(df.S1t, bins=100, histtype = 'step', density=True, label = 
```
I think here you should use `df2`, no? And `df1` for `S2t`

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    
+def monitor_S1(df, ebins, ns1bins, s1hbins, s1wbins):
+    fig, axs = plt.subplots(2, 2, figsize=(10, 8))
+    
+    nevents = len(df['event'].unique())
+    
+    df1 = df.groupby("event s2_peak".split()).first()
+    df1_ = df1.groupby('event').count()
+    df1__ = df1.groupby('event').mean()
+    
+    axs[0, 0].hist(df1_.nS1, ns1bins,
+                   density=True, histtype='step', label=
+                   f'run: {run_number}\n'
+                   f'events: {nevents}\n'
+                   f'mean: {df1_.nS1.mean():.2f}\n'
+                   f'std: {df1_.nS1.std():.2f}')
+    axs[0, 0].set_xlabel('Number of S1')
+    axs[0, 0].set_title('nS1 distribution')
+    axs[0, 0].grid(True)
+    axs[0, 0].legend()
```
This plot is already defined in the previous function, so perhaps the previous function is not necessary

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    
+    df1 = df.groupby("event s2_peak".split()).first()
+    df1_ = df1.groupby('event').count()
+    df1__ = df1.groupby('event').mean()
```
Not sure why you use the mean here

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    R2 = df.X[sel_DT]**2 + df.Y[sel_DT]**2
+    
+    DT = DT[sel].dropna()
+    R2 = R2[sel].dropna()
```
I think this won't work, because `sel` has a different length than `DT` and `R2`, but I don't see why not apply both selections at the same time.

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,531 @@
+    
+        
+    
+def monitor_S1_multipleruns(run_data, ebins, ns1bins, s1hbins, s1wbins):
```
I don't think we will use multiple runs here

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,157 @@
+dtrms2_low = lambda dt: -0.7 + 0.030 * (dt-20) # Gonzalo's
+dtrms2_upp = lambda dt: 2.6 + 0.036 * (dt-20) # Gonzalo's2
+dtrms2_cen = lambda dt:  1.0 + 0.033 * (dt-20)
+def dist_to_bandcenter(df): return df.Zrms**2 - dtrms2_cen(df.DT)
```
Is this function used somewhere?

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,157 @@
+
+
+dtrms2_low = lambda dt: -0.7 + 0.030 * (dt-20) # Gonzalo's
+dtrms2_upp = lambda dt: 2.6 + 0.036 * (dt-20) # Gonzalo's2
```
For now it's ok, but we should have these as inputs at some point

gonzaponte replied:
Now that these are in the config file, you can remove them.

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,157 @@
+
+
+def load_files(path):
+    return [str(p) for p in Path(path).rglob("*.h5")]
```
Probably not worth making it a function. We will see.

gonzaponte commented on invisible_cities/icaros/selection_functions.py:
```diff
@@ -0,0 +1,157 @@
+    col_name = 'Ec' if 'Ec' not in kdst.columns else 'Ec_2'
+    kdst[col_name] = corrected_energy.values
+
+    return kdst
```
This function should be in correcion_functions, I think.
Make `col_name` an argument of this function as well.

mcidlaso committed:
Remove monitoring functions

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,112 @@
+
+
+
+def select_lifetime_region(df, bins, x0, y0, shape, shape_size):
```
We will do this with symbols as well

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,112 @@
+        x1, y1 = x0-shape_size, y0-shape_size
+
+        mask_df = ((df.X >= x1) & (df.X <= x1+2*shape_size) & (df.Y >= y1) &
+                   (df.Y <= y1+2*shape_size))
```
`in_range`

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,112 @@
+def LT_fit(DT, s2e, p0):
+
+    f  = fit.fit(fit.expo, DT, s2e, p0)
+    return f.values, (f[2][0], f[2][1])
```
Probably not worth making a function just with this

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,94 @@
+    med_fun(empty_dst)
+
+
+def test_gaussianfit_does_fit_right():
```
```suggestion
def test_gaussianfit_empty():
```

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,94 @@
+    N = len(df_test)
+    result_med = med_fun(df_test)
+    assert result_med['nevents'].iloc[0] == N
+    assert fit_function(df_test, bins = 6)['nevents'].iloc[0] == N
```
```suggestion
    result_med = med_fun(df_test)
    result_fit = fit_function(df_test, bins = 6)
    assert result_med['nevents'].iloc[0] == N
    assert result_fit['nevents'].iloc[0] == N
```

gonzaponte commented on invisible_cities/icaros/krmap_functions_test.py:
```diff
@@ -0,0 +1,94 @@
+    
+
+
+def test_fitfun_medfun_Nevents():
+    d = {'DT': np.empty(6), 'x' : np.empty(6), 'y' : np.empty(6),'S2e' : np.ones(6)}
+    df_test = pd.DataFrame(data = d, index = range(0, 6))
+    N = len(df_test)
+    result_med = med_fun(df_test)
+    assert result_med['nevents'].iloc[0] == N
+    assert fit_function(df_test, bins = 6)['nevents'].iloc[0] == N
+
+
+def test_merge_maps():
+    Nan_map_test = create_empty_map(xy_range = (-500, 500), dt_range = (0, 1400), xy_nbins = 100, dt_nbins = 10)
+    d = {'nevents': np.empty(6, dtype = int), 'mu' : np.empty(6), 'sigma' : np.empty(6), 'mu_error' : np.empty(6), 'sigma_error' : np.empty(6)}
+    map_3D_test = pd.DataFrame(data = d, index = range(0, 6))
+    assert merge_maps(Nan_map_test, map_3D_test).shape == Nan_map_test.shape
```
`map_3D_test` needs to contain the columns `k`, `i`, and `j`.

mcidlaso committed:
Fix usage of dropna in normalization function

mcidlaso committed:
Use in_range to define xy region

mcidlaso committed:
Add NormMethod class to symbols.py

mcidlaso committed:
Import and use NormMethod class in normalization function

mcidlaso committed:
Correct error in gaussian_fit function

mcidlaso committed:
Use in_range function

mcidlaso committed:
Import new functions and create test for gauss seed

mcidlaso committed:
Create test for gauss seed

mcidlaso committed:
Redefined kr_map in normalization function so it doesn't contain nan values. Rewrote to use symbols and corrected some errors

mcidlaso committed:
Move apply_correctionmap function from selection_functions to correction_functions

mcidlaso committed:
Import functions and change test_method_norm to match the changes in normalization function

mcidlaso committed:
Fix mu value so it doesn't raise a warning when j = 0 (mu = 0) in test_apply_3D_map...()

mcidlaso committed:
Create test to check correction factor in apply_3Dmap function and to check the output shape in apply_correctionmap

mcidlaso committed:
Import more functions and change name in get_median() function from median and median_error to mu and mu_error

mcidlaso committed:
Change include_coordinates function so dt, x, y values are actually the bin centers

mcidlaso committed:
Change way of defining dtbins, xbins and ybins in compute_metadata

mcidlaso committed:
Change definitions of x_max, sigma and amp (area) in gauss_seed because it was not correct (still have to add some conditions?)

mcidlaso committed:
Rewrote conditions for using poisson_sigma() in quick_gauss_fit()

mcidlaso committed:
Use in_range in create_time_slices (are we going to move this function?) and create get_time_evol to compute time evolutions separately from append_time_evol (appends t_evols from different time slices)

mcidlaso committed:
Added tests for create_empty_map function and corrected get_median function tests

mcidlaso committed:
More corrections on get_median functions tests to match the changes done in get_median function

mcidlaso committed:
Added more tests for get_median and gaussian_fit functions (+ some corrections in test_gaussianfit_computes_right_values())

mcidlaso committed:
Added test for fit_map function

mcidlaso committed:
Test for merge_maps() function

mcidlaso committed:
Tests for include_coordinates function

mcidlaso committed:
Test for checking the columns of the output of compute_metadata

mcidlaso committed:
Test to check that quick_gauss_fit does the right fit

mcidlaso committed:
Test for create_time_slices function

mcidlaso committed:
More tests for get_time_evol and append_time_evol functions

mcidlaso committed:
Rewrote select_lifetime_region to use symbols. Removed LT_fit function since it wasn't really necessary

mcidlaso committed:
Changed compute_drift_v function to include nbins and dtrange in a single variable (dtbins)

mcidlaso committed:
Added functions to test select_lifetime_region for circle and square shapes

mcidlaso committed:
Moved apply_correctionmap to correction_functions.py, removed function that wasn't being used

mcidlaso committed:
Added class function to select region where to calculate lifetime

mcidlaso committed:
Redefine old variables df2 and df1 since there was no need to define new variables to groupby('event')

mcidlaso committed:
Use df2 and df1 instead of df

mcidlaso committed:
Change selection mask in monitor_kr_distribution to include DT and S2e selections in the same mask. Remove unnecessary functions

mcidlaso committed:
Rewrite plot_Ec function so the inputs are the column names that you want. Rewrite plot_lifetime_fit because LT_fit function no longer exists

mcidlaso committed:
Defined a new function to make selections in general, removed load_files function

mcidlaso committed:
Move sigmoid function from lifetime_vdrift_functions to invisible_cities/core/fit_functions

mcidlaso committed:
Correct errors in select_var_inrange

mcidlaso committed:
Move sigmoid function from lifetime_vdrift_functions to invisible_cities/core/fit_functions

mcidlaso committed:
Moved gauss_seed function to appear before in krmap_functions, modified some comments in get_median and name of arguments in gaussian_fit and fit_map

mcidlaso committed:
Modified comments in merge_maps, change name of arguments in compute_3D_map (bins to nbins_S2e, S2e_range)

mcidlaso committed:
Removed norm_method from metadata. Added xy_range, dt_range

mcidlaso committed:
Changed name of arguments in get_time_evol, removed p0 from args and fixed value inside the function updated append_time_evol to these changes

mcidlaso committed:
Now this assertion is not true because we changed sigma in get_median function as a function of len(df)

mcidlaso committed:
Change args names to match the changes done in gaussian_fit()

mcidlaso committed:
Removed norm_method and added xy_range, dt_range in test_metadata_column

mcidlaso committed:
Change tests for get_time_evol and append_time_evol to match the changes done to the functions

mcidlaso committed:
Add test for save_map function

mcidlaso committed:
Remove some comments and import sigmoid function from fit_functions in compute_drift_v

mcidlaso committed:
Add more tests to check the values inside the selected lifetime region (should I remove the limit values tests?). Also added tests for compute_drift_v() function

mcidlaso committed:
Added tests for sigmoid function in invisible_cities/core/fit_functions_test.py

mcidlaso committed:
Changed comment in select_1S1_1S2

mcidlaso committed:
Added documentation for functions and the type of each variable

mcidlaso committed:
Changed test_gauss_seed because we changed gauss_seed. Added missing tests and some corrections

mcidlaso committed:
Changed gauss_seed and len(df) limit in get_median

mcidlaso committed:
Corrected some errors in plotting functions

mcidlaso committed:
Removed function with repeated plots

mcidlaso committed:
Edited comment

mcidlaso committed:
Corrected test for normalization

mcidlaso committed:
Changed test for apply_3Dmap

mcidlaso committed:
Changed name and description to apply_correctionmap_inplace

mcidlaso committed:
Imported functions with correct name and added fixture functions for testing

mcidlaso committed:
Removed create_time_slices, edited get_time_evol to split dfs and get time evol at once

mcidlaso committed:
Included fixtures for testing krmap functions

mcidlaso committed:
Removed test for create time slices and changed tests for get time evol because the respective functions were removed or changed

mcidlaso committed:
Added fixtures for testing lifetime and vdrift functions

mcidlaso committed:
Change max to maximum in NormMethod so it doesnt confuse python

mcidlaso committed:
Import modules properly

mcidlaso committed:
Changes derived from renaming NormMethod.max to NormMethod.maximum

mcidlaso committed:
Changed get_median function so it works with any variable and not only with S2e. Had to change
the respective tests

mcidlaso committed:
Added first corrected energy to time evolution. Changed tests to match this change

mcidlaso committed:
Typo

mcidlaso committed:
Control plots

mcidlaso committed:
Added a function to make all plots at once and corrected some things

mcidlaso committed:
Modified control plots to remove verbosity

mcidlaso committed:
Add city and config

mcidlaso committed:
Forgot to add tests for selection functions

mcidlaso committed:
Rename city to Zemrude

mcidlaso committed:
Use file_out instead of name

mcidlaso committed:
Revisit file saving

mcidlaso committed:
Add monitoring plots functions

mcidlaso committed:
Changes to include efficiencies plot in control plots

mcidlaso committed:
removed matplotlib import

mcidlaso committed:
remove matplotlib imports

mcidlaso committed:
Forgot to include s2h in time evolution

mcidlaso committed:
Add canfranc monitoring plots to zemrude

mcidlaso committed:
Added argument in monitoring plots function to save them automatically

mcidlaso committed:
Add functions to apply 3D map to hits

mcidlaso committed:
Remove non necessary function

mcidlaso committed:
Avoid fit error when the slice in the lifetime fit region is empty

mcidlaso committed:
Added function to regularize maps (need to add tests for this function)
Added new density column to maps and test for this

mcidlaso committed:
Added minimum number of events to do the gaussian fit as a parameter.
If the number of events per bin is less than 5 now mu = nan.

mcidlaso committed:
Little change in apply_3Dmap and added rate of events in time evolution plots

mcidlaso committed:
Changes in krmap_functions_test to match changes done in get_median()

gonzaponte commented on invisible_cities/icaros/selection_functions_test.py:
```diff
@@ -0,0 +1,135 @@
+
+    assert np.all(kdst_sel.nS1.values == 1)
+    assert np.all(kdst_sel.nS2.values == 1)
+    assert np.all((kdst_sel.nS1.values) & (kdst_sel.nS2.values)) == 1
```
this line is probably not what you are trying to do. I'm surprised it even works.

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,93 @@
+
+        mask_df = in_range(df.X, x1, x1+2*shape_size) & in_range(df.Y, y1, y1+2*shape_size)
+
+        df_region = df[mask_df]
+        return df_region
+
+    if shape is SelRegionMethod.circle:
+
+        mask_df = (df.X - x0)**2 + (df.Y - y0)**2 <= shape_size**2
+        df_region = df[mask_df]
+        return df_region
+
```
```suggestion
    if shape is SelRegionMethod.square:
        x1, y1 = x0-shape_size, y0-shape_size
        mask_df = in_range(df.X, x1, x1+2*shape_size) & in_range(df.Y, y1, y1+2*shape_size)

    if shape is SelRegionMethod.circle:
        mask_df = (df.X - x0)**2 + (df.Y - y0)**2 <= shape_size**2

    df_region = df[mask_df]
    return df_region
```
Take the common part outside of the if statements

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    xy_range : tuple
+      Range in x and y (mm) inside which the map is being computed.
+    dt_range : tuple
+      Range in drift time (micro seconds) (same as z) inside which the map is being computed.
```
the units are not implicit, remove them.
also, strictly speaking, it's not the same as z, so that can also be misleading.

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    Returns
+    -------
+    Nan_map : pd.DataFrame
+      Dataframe filled with NaNs for each (i,j,k) combination.
```
i,j,k are not defined in the docstring, so better say something like "for each combination of z, x, y bins

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+def get_median(df        : pd.DataFrame,
+               nbins_S2e : int = None) -> pd.DataFrame:
+    """
+    Computes the median value of a given variable
+    -We consider that for len(df) < 3, it makes no sense to compute the std.
+    -Kr energy resolution is ~4%, so the std for S2e median would be ~0.04/2.35.
+    Parameters
```
update docstring

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+                                 'sigma_error': np.nan}, index = [0])
+
+    if not in_range(f.values[1], 0.5*df.S2e.median(), 1.5*df.S2e.median()):
+        return get_median(df)
```
document this behaviour in the docstring

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+            S2e_range     : tuple,
+            fit_function  : Callable,
+            min_events    : int = 50,
+            nbins_S2e     : int = None) -> pd.DataFrame:
```
let's discuss offline how to remove the optional arguments here.

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+            for j in range(1, mu.shape[2] -1):
+                if np.isnan(mu[k,i,j]):
+                    vals = [mu[k,i+ii,j+jj] for ii in (-1, 0, 1) for jj in (-1, 0, 1) if not (ii==jj==0)]
+                    if np.count_nonzero(np.isnan(vals))>3: continue
```
define a variable `max_nan_neighbours=3` at the top and use it here

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+                   nbins_S2e    : int = None) -> pd.DataFrame:
+    """
+    Computes a full 3D map merging (with merge_maps) a map full of nans from create_empty_map
+    with a 3D data map from fit_map and adding spatial coordinates using include_coordinates.
```
Avoid the implementation details. Give the overall picture. For instance:

Computes a 3D map from the data in `df` given a range in xy and dt and the number of bins for each coordinate. The map values in each voxel are computed using `fit_function`. A data quality filter is applied requiring S2e to be in a given range `S2e_range`.

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+
+    t_interval = tf_slice - ts
+
+    t_hours = t_interval/3600 #from seconds to hours
```
Rewrite this in a single line, and don't round to ints

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    coor_map = include_coordinates(map, xy_range, dt_range, xy_nbins, dt_nbins)
+
+    full_map = coor_map.assign(density = coor_map['nevents']/volume/t_hours)
+    #density in events*hour/cm^3
```
- put the comment above the line it applies to
-  it's events/hour/cm3
- bring the line that defines `volume` next to where it's used
- same for `t_hours`

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    tf_slice = int(df.time.values.max())
+
+    t_interval = tf_slice - ts
+
```
single line, no int

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    t_min = df.time.min()
+    t_max = df.time.max()
+
+    time_bins = np.arange(t_min, t_max + slice_seconds, slice_seconds)
```
`t_min` and `t_max` are a bit superfluous, use them directly

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,773 @@
+    Hdf file containing in each node each one of the inputs.
+    """
+
+    metadata.to_hdf(name, key = 'metadata', mode = 'a')
```
shouldn't it be `mode="w"`?

Explain why this is a special case with a comment

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,224 @@
+                  xy_params : dict = None) -> float:
+    """
+    Given a input krypton map, normalizes the whole map to a certain value
+    as a function of the chosen method.
```
It doesn't normalize the whole map. It returns the reference value which is later used for normalization

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,224 @@
+    if method is NormMethod.median_region_anode:
+        E_median_region_anode = anode.mu.median()
+        return E_median_region_anode
+
```
raise an error here. The only way to get to the end of this function if is `method` is not a `NormMethod`

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,224 @@
+                y            : pd.core.series.Series,
+                E            : pd.core.series.Series,
+                xy_params    : dict = None,
+                keV          : bool = False) -> pd.core.series.Series:
```
I think we discussed using an arbitraty unit instead of `keV`

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,224 @@
+
+
+
+def apply_correctionmap_inplace(kdst        : pd.DataFrame,
```
```suggestion
def apply_correctionmap_inplace_kdst(kdst        : pd.DataFrame,
```

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,293 @@
+from typing import Optional
+from typing import Tuple
+from typing import Union
+from .components import *
```
Replace `*` with the actual imports you need.
Tip: remove this line and run `pyflakes` on this file

mcidlaso committed:
Removed diffusion band definition since it is defined in the config file

mcidlaso committed:
Took the common part outside of the if statements

mcidlaso committed:
Implemented some changes in the documentation of the functions

mcidlaso committed:
Changed imports

mcidlaso committed:
Changed function name to apply_correctionmap_inplace_kdst

mcidlaso committed:
Changed tests to match function name change

mcidlaso committed:
ts is no longer previously defined

mcidlaso committed:
Changed imports

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,293 @@
+from .. icaros  .control_plots_functions import plot_time_evolution_with_errors_and_dates
+from .. types   .symbols              import NormMethod
+from .. types   .symbols              import SelRegionMethod
+from .. types   .symbols              import MapFitFunction
```
align and sort IC imports

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,293 @@
+
+
+def concatenated_dsts_from_files(path: List[str], group: str, node:str)-> Iterator[Dict[str,Union[pd.DataFrame, int, np.ndarray]]]:
+
```
remove this blank line

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,293 @@
+
+def create_selfmap(xy_range, dt_range, xy_nbins, dt_nbins, S2e_range, fit_function, nbins, min_events):
+    if fit_function == MapFitFunction.gaussian:
+        fit_function = gaussian_fit_ready
```
This should be
```suggestion
        fit_function = gaussian_fit_ready(nbins, min_events)
```
right?

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,293 @@
+                                          nbins)
+    else:
+        fit_function = get_median
+
```
This is already done in `create_selfmap`

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,287 @@
+    apply_3Dmap_to_data = fl.map( apply_selfmap(norm_method,
+                                                xy_params,
+                                                'Ec_2',
+                                                 unit = keV)
```
4th argument is misaligned

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,287 @@
+                                    s1_DTrange,
+                                    bins_Ec,
+                                    error
+                                    )
```
revisit alignment

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,287 @@
+                                                 shape_size,
+                                                 xy_range_plot,
+                                                 names
+                                                 )
```
revisit alignment

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,287 @@
+
+                                           fl.fork(save_everything,
+                                                   make_control_plots,
+                                                   plot_t_evol))
```
revisit alignment

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,287 @@
+                          , args = ('time_evol'))
+
+
+    with tb.open_file(file_out, "w", filters=tbl.filters(compression)):
+        pass
```
is this to overwrite the output file, if it exists and set the compression? If so, explain it in a comment

gonzaponte commented on invisible_cities/core/fit_functions_test.py:
```diff
@@ -252,6 +252,34 @@ def test_power_at_zero(a, b):
+
+    assert y.shape == x.shape
+
+
```
Remove comments

gonzaponte commented on invisible_cities/icaros/control_plots_functions.py:
```diff
@@ -0,0 +1,782 @@
+from .. core.fit_functions import sigmoid
+from .. core.core_functions import in_range
+from .. core.core_functions import shift_to_bin_centers
+from .. icaros.lifetime_vdrift_functions import select_lifetime_region
```
sort and align imports

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,219 @@
+                dt          : pd.core.series.Series,
+                x           : pd.core.series.Series,
+                y           : pd.core.series.Series,
+                E           : pd.core.series.Series,
```
change `pd.Series` to `np.ndarray`

gonzaponte commented on invisible_cities/icaros/correction_functions.py:
```diff
@@ -0,0 +1,219 @@
+                y           : pd.core.series.Series,
+                E           : pd.core.series.Series,
+                xy_params   : dict  = None,
+                unit        : Union[float, NoneType]  = keV) -> pd.core.series.Series:
```
```suggestion
                xy_params   : Union[dict, NoneType]  = None,
                unit        : Union[float, NoneType]  = keV) -> np.ndarray:
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,722 @@
+"""
+This script contains functions to compute empty maps, 3D maps from kdsts and merge them to create an 'uniform' map.
+It also contains functions to get and store map info (metadata) and time evolution.
+"""
```
This goes at the very beginning of the file, before the imports.

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,722 @@
+    return Nan_map
+
+
+def get_median(var : pd.core.series.Series) -> pd.DataFrame:
```
```suggestion
def get_median(var : np.ndarray) -> pd.DataFrame:
```

gonzaponte commented on invisible_cities/icaros/krmap_functions.py:
```diff
@@ -0,0 +1,722 @@
+
+
+
+def gaussian_fit(var        : pd.core.series.Series,
```
```suggestion
def gaussian_fit(var        : np.ndarray,
```

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,91 @@
+import invisible_cities.core.fit_functions as fit
+from invisible_cities.core.core_functions import in_range, shift_to_bin_centers
+from invisible_cities.types.symbols import SelRegionMethod
+
```
organize and align the imports

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions.py:
```diff
@@ -0,0 +1,91 @@
+    if shape is SelRegionMethod.circle:
+
+        mask_df = (df.X - x0)**2 + (df.Y - y0)**2 <= shape_size**2
+
```
remove spurious blank lines

gonzaponte commented on invisible_cities/icaros/lifetime_vdrift_functions_test.py:
```diff
@@ -0,0 +1,96 @@
+from .. core   .core_functions import fix_random_seed
+from .. types  .symbols        import SelRegionMethod
+from .. icaros .lifetime_vdrift_functions import select_lifetime_region
+from .. icaros .lifetime_vdrift_functions import compute_drift_v
```
align imports

gonzaponte reviewed (commented):
A few cosmetic changes and we are done. We will rewrite the history together to reduce the number of commits.

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,275 @@
+
+                       fl.fork(save_everything,
+                               make_control_plots,
+                               plot_t_evol))
```
```suggestion
            ,pipe   = fl.pipe(apply_preliminary_map,
                               apply_selections,
                               compute_3D_map,
                               compute_metadata
                               apply_3Dmap_to_data,
                               get_t_evol,

                               fl.fork(save_everything,
                                       make_control_plots,
                                       plot_t_evol))
```

gonzaponte commented on invisible_cities/cities/zemrude.py:
```diff
@@ -0,0 +1,275 @@
+
+def select_dst(dtrms2_low, dtrms2_upp, low_xrays, high_xrays, low_S2t, high_S2t, R_max, low_DT, high_DT, low_nsipm, high_nsipm):
+    def apply_selections_dst(df):
+        return apply_selections(df, dtrms2_low, dtrms2_upp, low_xrays, high_xrays, low_S2t, high_S2t, R_max, low_DT, high_DT, low_nsipm, high_nsipm)
+    return apply_selections_dst
+
+
+def create_selfmap(xy_range, dt_range, xy_nbins, dt_nbins, S2e_range, fit_function, nbins, min_events):
+    if fit_function == MapFitFunction.gaussian:
+        fit_function = gaussian_fit_ready(nbins, min_events)
+    else:
+        fit_function = get_median
+    def create_map(df):
```
Let's add a protection here:

```suggestion
    if fit_function == MapFitFunction.gaussian:
        fit_function = gaussian_fit_ready(nbins, min_events)
    elif fit_function == MapFitFunction.median:
        fit_function = get_median
    else:
        raise ValueError(f"Invalid fit function {fit_function}")

    def create_map(df):
```

mcidlaso committed:
Add sigmoid function

mcidlaso committed:
Add selection functions

mcidlaso committed:
Add functions to compute lifetime and drift velocity

mcidlaso committed:
Add functions to create krypton maps

mcidlaso committed:
Add functions to apply krypton maps to data

mcidlaso committed:
Add functions to plot control plots

mcidlaso committed:
Add zemrude city

mcidlaso committed:
Add 3D map to test_data

mcidlaso committed:
Add function to load maps

mcidlaso committed:
Remove incorrect assertion in test

Since esmeralda performs energy corrections, there is no guarantee
that the energy before and after esmeralda is the same

mcidlaso committed:
Rename normalization-related parameters

mcidlaso committed:
Cosmetics

mcidlaso committed:
Update fixtures

mcidlaso committed:
Update city configs

mcidlaso committed:
Use 3D map in hits corrector

mcidlaso committed:
Update reference test files

mcidlaso committed:
Update tests for hits corrector

mcidlaso committed:
Skip or remove old normalization tests

Skip some tests that were using temporal correction maps. Remove test
for normalization strategies since we already have tests for
normalization methods.

mcidlaso committed:
Apply z to DT correction

mcidlaso committed:
Apply temporal correction

mcidlaso committed:
Correct typos

mcidlaso committed:
Move apply_correctionmap function

The map is defined using x,y,dt, so we need to convert DT to Z
afterwards.

mcidlaso committed:
Add test data for Zemrude

mcidlaso committed:
Update zemrude config

mcidlaso committed:
Add test for zemrude city

gonzaponte reviewed (approved):
The 3D version of our beloved Kr maps. An important addition developed with peras, reasonably tested and validated externally. Good job!

mcidlaso committed:
Remove selected_dst from control plots.

Since apply correction map function is done in place, df_selected and
df_corr end up having the same Ec and Ec2 columns

mcidlaso committed:
Remove spurious column in krypton map
